### PR TITLE
[Serializer] Add Support of recursive denormalization on object_to_populate

### PR DESCRIPTION
**Pull request by jewome62**
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | Pending
| Fixed tickets | #21669
| License       | MIT
| Doc PR        | Pending

Currently the deserialization re-create new sub-object with object_to_populate.
This option permit to make object_to_populate recursive.

### CODE REVIEW

**Comment by joelwurtz** [comment]
Nice feature, does this works on collection ? (don't think so seeing the code, and wondering if it's doable)

**Comment by teohhanhui** [comment]
See https://github.com/symfony/symfony/issues/21669#issuecomment-280889824

I don't think those issues are addressed.

**Comment by jewome62** [comment]
@joelwurtz i test it tomorrow

**Comment by jewome62** [comment]
@joelwurtz you have right, an object collection in the parent has no populated object, the problem is to have a reference for linked the object passed in the 'object_to_populate' and the object received by decoding.

Either we consider this PR for objects containing other objects, or we have a solution to bind collection objects to the input received

**Comment by joelwurtz** [comment]
I don't think this PR should solve that problem, futhermore, some users would have different behavior:

 * Replace: like actually
 * Merge: mainly if the index key is the same as the populated object we use the previous object for as an object to populate; also if previous index is not present we remove it, and if the new index is not present in the old list we add it
 * Append: keep current lists and add new object to the list

**Review by joelwurtz — APPROVED** [review]

File: src/Symfony/Component/Serializer/Normalizer/ObjectToPopulateTrait.php [hunk 1]
```diff
@@ -26,7 +26,7 @@ trait ObjectToPopulateTrait
      */
     protected function extractObjectToPopulate($class, array $context, $key = null)
     {
-        $key = $key ?: 'object_to_populate';
+        $key = $key ?: AbstractNormalizer::OBJECT_TO_POPULATE;
```
Comment (dunglas) [inline src/Symfony/Component/Serializer/Normalizer/ObjectToPopulateTrait.php]: `??` ?

File: src/Symfony/Component/Serializer/CHANGELOG.md [hunk 2]
```diff
@@ -5,6 +5,7 @@ CHANGELOG
 -----
 
  * added the list of constraint violations' parameters in `ConstraintViolationListNormalizer`
+ * added a `deep_object_to_populate` context option to recursive denormalize on `object_to_populate` object.
```
Comment (dunglas) [inline src/Symfony/Component/Serializer/CHANGELOG.md]: Cannot it be the default?

Comment (fabpot) [reply src/Symfony/Component/Serializer/CHANGELOG.md]: Would it be a BC break?

Comment (jewome62) [reply src/Symfony/Component/Serializer/CHANGELOG.md]: The current behavior is to not populate the children object. So if we enable it, this is going to be a BC break.
So I think we could enable it by default in Symfony 5.

Comment (fabpot) [reply src/Symfony/Component/Serializer/CHANGELOG.md]: Cannot be change in 5 either as we need an upgrade path from 4 to 5, which is going to be impossible.

**Comment by fabpot** [comment]
@jewome62 You need to rebase the PR as we don't merge PRs with merge commits.

**Comment by jewome62** [comment]
> @jewome62 You need to rebase the PR as we don't merge PRs with merge commits.

I fix that, I have use the confilct resolver into from github

**Comment by jewome62** [comment]
Failure of unit tests is not related to this PR

**Comment by fabpot** [comment]
Thank you @jewome62.

File: src/Symfony/Component/Serializer/Normalizer/AbstractObjectNormalizer.php [hunk 3]
```diff
@@ -38,6 +39,7 @@ abstract class AbstractObjectNormalizer extends AbstractNormalizer
     const SKIP_NULL_VALUES = 'skip_null_values';
     const MAX_DEPTH_HANDLER = 'max_depth_handler';
     const EXCLUDE_FROM_CACHE_KEY = 'exclude_from_cache_key';
+    const DEEP_OBJECT_TO_POPULATE = 'deep_object_to_populate';
```
Comment (dbu) [inline src/Symfony/Component/Serializer/Normalizer/AbstractObjectNormalizer.php]: in #30888 i am adding phpdoc to these configuration constants to explain what they do and what values they accept.

File: src/Symfony/Component/Serializer/Normalizer/AbstractObjectNormalizer.php [hunk 4]
```diff
@@ -274,6 +276,13 @@ public function denormalize($data, $class, $format = null, array $context = [])
                 continue;
             }
 
+            if ($context[self::DEEP_OBJECT_TO_POPULATE] ?? $this->defaultContext[self::DEEP_OBJECT_TO_POPULATE] ?? false) {
+                try {
+                    $context[self::OBJECT_TO_POPULATE] = $this->getAttributeValue($object, $attribute, $format, $context);
```
Comment (dbu) [inline src/Symfony/Component/Serializer/Normalizer/AbstractObjectNormalizer.php]: this overwrites the context. i think this means that if there are more loops, you might have a previous object to populate in the context. i think we should unset OBJECT_TO_POPULATE if it does not find anything and when the DEEP_OBJECT_TO_POPULATE flag is not true.

Comment (dbu) [reply src/Symfony/Component/Serializer/Normalizer/AbstractObjectNormalizer.php]: we found where OBJECT_TO_POPULATE is unset. so this is fine as is.

i will try to do a test and fix to make OBJECT_TO_POPULATE more robust, as i think there are edge cases that would lead to weird behaviour.

File: src/Symfony/Component/Serializer/Normalizer/AbstractObjectNormalizer.php (hunk 3 above)
Comment (dbu) [reply src/Symfony/Component/Serializer/Normalizer/AbstractObjectNormalizer.php]: as this is already merged, i will rebase #30888 and document it there. and extract the test for this into a trait like we do the other tests there.

File: src/Symfony/Component/Serializer/Normalizer/AbstractObjectNormalizer.php (hunk 4 above)
Comment (dbu) [reply src/Symfony/Component/Serializer/Normalizer/AbstractObjectNormalizer.php]: => #30977

**Comment by dbu** [comment]
documentation PR: https://github.com/symfony/symfony-docs/pull/11344
